### ISSUE TITLE
Redundant error message

### ISSUE THREAD

**Issue by HolQue** [test-fullautomation/robotframework-testsuitesmanagement#67]
A not existing configuration causes two error messages with the mostly same content:
```
[ ERROR ] The local configuration file C:/RobotTest/localconfig/iamnotexisting.json which set in ROBOT_LOCAL_CONFIG variable, does not exist!!!
[ ERROR ] Calling method '_start_suite' of listener 'LibListener' failed: Unknown exception occurs. Details: The local configuration file C:/RobotTest/localconfig/iamnotexisting.json does not exist!!!

```
I would prefer to have the second one removed, because this second error message does not provide any additional useful information for the user.

Reference:
https://github.com/test-fullautomation/robotframework-testsuitesmanagement/tree/HolQue/task/documentation_updates_and_fixes

**Comment by namsonx** [comment]
Hello Holger,

    The second error message comes from LibListener - the event listener of Robot Framework that I'm using to define the activities before started testsuite.
    To remove this second error message, we have to modify core of Robot FW, and I think this modification will not make sense for Robot FW community.

Thank you,
Son

**Comment by HolQue** [comment]
I still have doubts. The LibListener.py is part of the RobotFramework_TestsuitesManagement sources and contains the following code, that causes the redundant log entries:

```
logger.error(f"The local configuration file {os.environ['ROBOT_LOCAL_CONFIG']} which set in ROBOT_LOCAL_CONFIG variable, does not exist!!!")
BuiltIn().unknown(f"The local configuration file {os.environ['ROBOT_LOCAL_CONFIG']} does not exist!!!")
```

For what purpose do you need to call `BuiltIn().unknown()`?

By the way: The test result is passed. But why do you continue with test execution? In case of such kind of errors I would expect to stop the test execution.

**Comment by namsonx** [comment]
Hello Holger,

Robot FW allows users override _start_suite method in case we want to define some activities at the very beginning of test suite. And I defined LibListener.py as part of the RobotFramework_TestsuitesManagement sources to override _start_suite method.

For the purpose of BuiltIn().unknown(), we call this method to set the test results are unknown when the Robot execution failed due to issues from setup environment, config, ... not form software under test.

Thank you,
Son

**Comment by HolQue** [comment]
But like I mentioned in my last comment: The test result is not unknown! Even in case of such a local config error the test is executed and the result is passed. It looks like that the BuildIn().unknown() has no effect. I still do not understand why we need both a logger.error() and a BuildIn().unknown(). Is there no possibility to synchronize this in a better way to avoid such redundancies?

**Comment by namsonx** [comment]
> But like I mentioned in my last comment: The test result is not unknown! Even in case of such a local config error the test is executed and the result is passed. It looks like that the BuildIn().unknown() has no effect. I still do not understand why we need both a logger.error() and a BuildIn().unknown(). Is there no possibility to synchronize this in a better way to avoid such redundancies?

Oh, I will check this issue. Thank you for your finding

Thank you,
Son

**Comment by namsonx** [comment]
Hello Holger,

I have just pushed the fix for this issue, I also cherry-picked to your feature branch HolQue/task/documentation_updates_and_fixes

The error log message and the test results now look like below:

[ ERROR ] The local configuration file C:\RobotTest\localconfig\local_config.json which set in ROBOT_LOCAL_CONFIG variable, does not exist!!!
Test builtin Variable with Config level 1                             | UNKNOWN |
1.332 != 1.999

Thank you,
Son

**Comment by HolQue** [comment]
Hello Son,

thank you very much for the update. Looks much better now in console! The error message also appears properly in xml and html output files of the Robot Framework - but unfortunately not in the text log file created by the -b command line parameter. This is really a big disadvantage, because this log file in text format is the only one that is really useful for debugging.

@ThPoll 
This is in general an open point. We need concept that determines how to handle error messages:
* Which function
* under which conditions
* writes what kind of detail in which output file

Maybe for all output a logging wrapper needs to be introduced - similar to the CLogging class in TML.

Please create an investigation and concept task for this issue.

**Comment by namsonx** [comment]
Hello Holger,

I just created [new commit](https://github.com/test-fullautomation/robotframework-testsuitesmanagement/pull/63/commits/787037f66990306f3a9abf32d3d0510ff5f16f3b) to interpret the reason of Unknow status when loading configuration file failed.

For the debug log you mentioned in above comment, I think we can investigate and implement for next release :)

Thank you,
Son

**Comment by HolQue** [comment]
Hello Son,

yes - next release! :-)

**Comment by HolQue** [comment]
I tried to get more clarity - therefore some additional thoughts here about the logging:

We have two methods with different functionality:

`logger.error(<error message>)` and `BuiltIn().unknown(<error message>)`

And the first one prints the error message starting with red [ERROR] string. The second one makes the test result UNKNOWN and also provides the possibility to define an error message.

We are in situations where we need to call both functions, but we do not want to have the error message printed twice. Therefore only one of the function calls should contain this error message - maybe the `logger.error()`. Then `BuiltIn().unknown()` is called without error message.

Unfortunately this does not work properly.

Because `logger.error(<error message>) `does not write the error message into the text log file (-b command line option) - but `BuiltIn().unknown(<error message>)` does. We do not have the error message twice any more in console, but we also do not have the error message in text log file any more. In my opinion this is not acceptable.

And also `BuiltIn().unknown()` has it's own issue: Already several times it was annoying for me to see this message in the text log file:
   `UNKNOWN - Unknown exception occurs. Details: None`
An exception without any details about it? Sounds worryingly for me.

In RobotFramework\python39\Lib\site-packages\robotframework-4.1.3-py3.9.egg\robot\errors.py I found the reason:
```
   def _get_message(self):
       msg = "Unknown exception occurs. Details: %s" % self.msg
       return msg
```
With `self.msg` initialized with `None`.

I do not like this implementation! The default value makes the error message an option! And in case of a user decides not to provide an error message when he uses `BuiltIn().unknown()` - for what reason ever - then the function has to be quiet! Then the function has to set the test result to UNKNOWN, nothing else.

In my opinion the `self.msg` has to be checked against `None` and "". And only in case of a user provides something, the function is allowed to add "`Unknown exception occurs. Details: `" to the message. Otherwise `self.msg` should be returned unchanged.

But this also does not solve the problem mentioned above. We need a strict separation of concern:
* A function only for printing an error message, but to every possible destination (console and all log files) and without doing anything else
* A function influencing the test result (with maybe printing also an error message, but only as option)
